File: client/app/src/main/java/com/rh/heji/data/db/ImageDao.java
package com.rh.heji.data.db;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import java.util.List;

/**
 * Date: 2020/11/19
 * Author: 锅得铁
 * #
 */
@Dao
public interface ImageDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void install(Image ticket);

    @Insert(entity = Image.class,onConflict = OnConflictStrategy.REPLACE)
    void install(List<Image> ticket);

    @Delete
    void delete(Image ticket);

    @Query("delete from bill_img where _bid =:billID")
    void deleteBillImage(String billID);

    @Transaction
    @Query("update bill_img set img_path=:imagePath, sync_status =:status where _id =:id")
    int updateImageLocalPath(String id, String imagePath, int status);

    @Transaction
    @Query("update bill_img set img_online_path=:onlinePath, sync_status=:status  where _id =:imgId")
    int updateOnlinePath(Long imgId, String onlinePath, int status);

    @Query("select * from bill_img where img_online_path=:path")
    List<Image> findByOnLinePath(String path);

    @Query("DELETE FROM " + Image.TAB_NAME + " WHERE " + Image.COLUMN_ID + "=:imgID")
    void deleteById(String imgID);

    @Query("SELECT * FROM bill_img WHERE _bid =:billId")
    List<Image> findByBillId(String billId);

    @Query("SELECT * FROM bill_img WHERE _id =:id")
    List<Image> findById(String id);

    @Query("SELECT * FROM bill_img WHERE _bid =:billId AND sync_status==" + Constant.STATUS_NOT_SYNC)
    List<Image> findByBillIdNotAsync(String billId);

    @Query("SELECT * FROM bill_img WHERE (img_path ISNULL OR img_path=='') AND(img_online_path!='' OR img_online_path != NULL)")
    LiveData<List<Image>> observerNotDownloadImages();
}

File: client/app/src/main/java/com/rh/heji/data/db/ImageDao.kt
package com.rh.heji.data.db

import androidx.room.*
import com.rh.heji.App
import kotlinx.coroutines.flow.Flow

/**
 * Date: 2020/11/19
 * @author: 锅得铁
 * #
 */
@Dao
interface ImageDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun install(ticket: Image )

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun install(ticket: MutableList<Image >)

    @Transaction
    fun installBillAndImages(bill:Bill,image:MutableList<Image>):Long{
        var count =App.dataBase.billDao().install(bill)
        install(image)
        return count
    }

    @Query("SELECT * FROM image WHERE image_id IN (:img_ids) AND sync_status !=${STATUS.DELETED}")
    fun findImage(img_ids:List<String>):MutableList<Image>


    @Query("SELECT * FROM image WHERE bill_id =:billID AND sync_status =:status")
    fun findByBillID(billID:String,status:Int):MutableList<Image>

    @Query("DELETE FROM image WHERE bill_id =:billID")
    fun deleteBillImage(billID: String )

    @Query("SELECT image_id FROM image WHERE bill_id =:billID AND sync_status !=${STATUS.DELETED}")
    fun findImagesId(billID: String ):MutableList<String>

    @Transaction
    @Query("UPDATE image SET sync_status = ${STATUS.DELETED} WHERE image_id =:imageID")
    fun preDelete(imageID: String )

    @Transaction
    @Query("UPDATE image SET local_path=:imagePath, sync_status =:status WHERE image_id =:id")
    fun updateImageLocalPath(id: String , imagePath: String , status: Int): Int

    @Transaction
    @Query("UPDATE image SET online_path=:onlinePath, sync_status=:status  WHERE image_id =:imgId")
    fun updateOnlinePath(imgId: String , onlinePath: String , status: Int): Int

    @Query("SELECT * FROM image WHERE online_path=:path")
    fun findByOnLinePath(path: String ): MutableList<Image >

    @Query("DELETE FROM ${Image.TAB_NAME} WHERE ${Image.COLUMN_ID}=:imgID")
    fun deleteById(imgID: String )

    @Query("SELECT * FROM image WHERE bill_id =:billId AND sync_status !=${STATUS.DELETED}")
    fun findByBillId(billId: String ): Flow<MutableList<Image>>

    @Query("SELECT * FROM image WHERE (local_path ISNULL OR local_path='') AND(online_path!='' OR online_path != NULL)")
    fun observerNotDownloadImages(): Flow<MutableList<Image>>

}